### PR TITLE
Fix many of the graphics artifacts while building draggable reservoirs

### PR DESCRIPTION
**Pull request by crudelios**
This change is specifically tailored to draggable reservoirs (where you build two reservoirs and an aqueduct simply by click-dragging).

The following changes were implemented:

- Makes aqueducts not turn towards every tile of the reservoir construction ghost, but only to those tiles that are aqueduct entrances
- Makes reservoir ghosts properly draw behind/in front of the other reservoir ghost
- Makes aqueducts show in front of the reservoir ghost
- Better display when reservoirs have water access: If one of the reservoirs is being built close to water, both will show as being filled

It's not perfect: some artifacts still exist regarding aqueducts, but they are less noticeable. Fixing those would require too many changes to the game for little difference, so I don't think it's worth the extra effort.

### CODE REVIEW

**Comment by bvschaik** [comment]
General observations:
- The cost of the reservoir combination now stays fixed on the starting reservoir. That's not handy when your first reservoir scrolls off the viewport.
- I'm not sure I prefer the new drawing order, since it leads to these kind of "broken aqueduct" artifacts:
broken-aqueduct

I think changing just the "Makes aqueducts not turn towards every tile of the reservoir construction ghost, but only to those tiles that are aqueduct entrances" bit will improve things a lot already, without leading to further artifacts.

**Comment by crudelios** [comment]
> The cost of the reservoir combination now stays fixed on the starting reservoir. That's not handy when your first reservoir scrolls off the viewport.

I'm not sure what's causing that, but I'll ckeck.

I think "Makes reservoir ghosts properly draw behind/in front of the other reservoir ghost" can also be kept without creating new artifacts. I'll check.

**Comment by crudelios** [comment]
I don't know if you're notified when I commit the requested code changes, so I'm posting this to let you know.

**Comment by CommonLoon102** [comment]
> 
> 
> I don't know if you're notified when I commit the requested code changes, so I'm posting this to let you know.

You can ask for a code review here on the top right.

**Comment by crudelios** [comment]
> You can ask for a code review here on the top right.

Never thought about that. Thanks.

**Comment by bvschaik** [comment]
> I don't know if you're notified when I commit the requested code changes, so I'm posting this to let you know.

I do get notified for new commits, both by email and on Github. I just haven't had time to look at them yet :)

File: src/widget/city_building_ghost.c [hunk 1]
```diff
@@ -293,16 +305,29 @@ static void draw_draggable_reservoir(const map_tile *tile, int x, int y)
...
+            int map_y_start = map_grid_offset_to_y(offset) - 1;
+            if (!has_water) {
+                has_water = map_terrain_exists_tile_in_area_with_type(map_x_start - 1, map_y_start - 1, 5, TERRAIN_WATER);
+            }
+            if (map_x_start > map_x || map_y_start > map_y) {
+                draw_later = 1;
+            } else {
+                draw_single_reservoir(x_start, y_start, has_water);
+            }
+            city_view_pixels_to_grid_offset(x, y);
```
Comment (bvschaik) [inline src/widget/city_building_ghost.c]: Is this line still necessary?

File: src/widget/city_building_ghost.c [hunk 2]
```diff
@@ -293,16 +305,29 @@ static void draw_draggable_reservoir(const map_tile *tile, int x, int y)
...
             }
         } else {
-            draw_building(image_group(GROUP_BUILDING_RESERVOIR), x_start, y_start);
+            offset = city_view_pixels_to_grid_offset(x_start, y_start);
+            int map_x_start = map_grid_offset_to_x(offset) - 1;
+            int map_y_start = map_grid_offset_to_y(offset) - 1;
+            if (!has_water) {
+                has_water = map_terrain_exists_tile_in_area_with_type(map_x_start - 1, map_y_start - 1, 5, TERRAIN_WATER);
+            }
+            if (map_x_start > map_x || map_y_start > map_y) {
```
Comment (bvschaik) [inline src/widget/city_building_ghost.c]: This check does not result in correct drawing order if the map is rotated.

**Pull request merged**